# Movie Review Dataset
- Cornell movie review dataset
- https://www.cs.cornell.edu/people/pabo/movie-review-data/
- pos 5331, neg 5331, total 10662

folder_name\
&emsp;ㄴ data\
&emsp;&emsp;ㄴ rt-polarity.neg\
&emsp;&emsp;ㄴ rt-polarity.pos\
&emsp;ㄴ lec_02.ipynb

In [1]:
# from google.colab import drive
# drive.mount('/content/drive')

In [4]:
pos_review = []
neg_review = []

with open("rt-polarity.pos", 'r', encoding='latin-1') as rf:
    pos_review = [line.strip() for line in rf.readlines()]

with open("rt-polarity.neg", 'r', encoding='latin-1') as rf:
    neg_review = [line.strip() for line in rf.readlines()]

dataset = pos_review + neg_review

In [5]:
len(dataset), dataset[:10]

(10662,
 ['the rock is destined to be the 21st century\'s new " conan " and that he\'s going to make a splash even greater than arnold schwarzenegger , jean-claud van damme or steven segal .',
  'the gorgeously elaborate continuation of " the lord of the rings " trilogy is so huge that a column of words cannot adequately describe co-writer/director peter jackson\'s expanded vision of j . r . r . tolkien\'s middle-earth .',
  'effective but too-tepid biopic',
  'if you sometimes like to go to the movies to have fun , wasabi is a good place to start .',
  "emerges as something rare , an issue movie that's so honest and keenly observed that it doesn't feel like one .",
  'the film provides some great insight into the neurotic mindset of all comics -- even those who have reached the absolute top of the game .',
  'offers that rare combination of entertainment and education .',
  'perhaps no picture ever made has more literally showed that the road to hell is paved with good intentions .',


# Construct Vocab(word level) with Text Preprocessing

In [6]:
from tqdm import tqdm
from nltk.tokenize import word_tokenize

In [7]:
import nltk

nltk.download("punkt")

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [8]:
def construct_vocab(dataset):
    ## start ##
    vocab = {}
    for sentence in dataset:
        words = word_tokenize(sentence)
        for word in words:
            vocab[word] = vocab.get(word, 0) + 1
    return vocab

def vocab_analysis(vocab):
    print("Vocab size: ", len(vocab))
    print("Vocab sample: ", list(vocab.items())[:10])

### 0. None

In [9]:
vocab_none = construct_vocab(dataset)

vocab_analysis(vocab_none)

Vocab size:  20298
Vocab sample:  [('the', 10113), ('rock', 34), ('is', 3559), ('destined', 8), ('to', 4234), ('be', 939), ('21st', 6), ('century', 18), ("'s", 3537), ('new', 206)]


### 1. Lower case

In [10]:
def lowercase(dataset):
    ## start ##
    ret = [sentence.lower() for sentence in dataset]

    ### end ###

    return ret

In [11]:
dataset_lowercase = lowercase(dataset)
vocab_lowercase = construct_vocab(dataset_lowercase)

vocab_analysis(vocab_lowercase)

Vocab size:  20297
Vocab sample:  [('the', 10113), ('rock', 34), ('is', 3559), ('destined', 8), ('to', 4234), ('be', 939), ('21st', 6), ('century', 18), ("'s", 3537), ('new', 206)]


### 2. Stopword Removal

In [12]:
import nltk
from nltk.stem import PorterStemmer
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet
from nltk.corpus import stopwords

nltk.download("punkt")
nltk.download('wordnet')
nltk.download('averaged_perceptron_tagger')
nltk.download('stopwords')

example_words = ["plays","played","play","playing"]
test_sentence = [
    "I played and started playing with players in the playgrounds",
    "I am playing with my playmates"
    ]

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [13]:
stop_words = set(stopwords.words('english'))  # 영어 불용어 리스트 가져오기

def remove_stopwords(dataset):
    global stop_words

    ## start ##
    new_dataset = [' '.join([word for word in word_tokenize(sentence) if word.lower() not in stop_words]) for sentence in dataset]

    ### end ###

    return new_dataset

print("Before stopword removal: ", test_sentence)
print("After stopword removal: ", remove_stopwords(test_sentence))

Before stopword removal:  ['I played and started playing with players in the playgrounds', 'I am playing with my playmates']
After stopword removal:  ['played started playing players playgrounds', 'playing playmates']


In [14]:
dataset_remove_stopwords = remove_stopwords(dataset)
vocab_remove_stopwords = construct_vocab(dataset_remove_stopwords)

vocab_analysis(vocab_remove_stopwords)

Vocab size:  20165
Vocab sample:  [('rock', 34), ('destined', 8), ('21st', 6), ('century', 18), ("'s", 3537), ('new', 206), ('``', 657), ('conan', 2), ('going', 86), ('make', 278)]


### 3. Stemming

In [16]:
ps = PorterStemmer()

print("{0:20}{1:20}".format("--Word--","--Stem--"))
for word in example_words:
   print ("{0:20}{1:20}".format(word, ps.stem(word)))

--Word--            --Stem--            
plays               play                
played              play                
play                play                
playing             play                


In [17]:
def stemming(dataset):
    global ps

    ## start ##
    new_dataset = [' '.join([ps.stem(word) for word in word_tokenize(sentence)]) for sentence in dataset]
    ### end ###

    return new_dataset

print("Before stemming: ", test_sentence)
print("After stemming: ", stemming(test_sentence))

Before stemming:  ['I played and started playing with players in the playgrounds', 'I am playing with my playmates']
After stemming:  ['i play and start play with player in the playground', 'i am play with my playmat']


In [18]:
dataset_stemming = stemming(dataset)
vocab_stemming = construct_vocab(dataset_stemming)

vocab_analysis(vocab_stemming)

Vocab size:  14873
Vocab sample:  [('the', 10113), ('rock', 37), ('is', 3559), ('destin', 19), ('to', 4234), ('be', 1108), ('21st', 6), ('centuri', 22), ("'s", 3537), ('new', 207)]


### 4. Lemmatization

In [19]:
def get_wordnet_pos(word):
    tag = nltk.pos_tag([word])[0][1][0].upper()
    tag_dict = {"J": wordnet.ADJ, "N": wordnet.NOUN, "V": wordnet.VERB, "R": wordnet.ADV}
    return tag_dict.get(tag, wordnet.NOUN)

In [20]:
nltk.download('averaged_perceptron_tagger_eng')

[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger_eng is already up-to-
[nltk_data]       date!


True

In [21]:
le = WordNetLemmatizer()

print("{0:20}{1:20}".format("--Word--","--Stem--"))
for word in example_words:
   print ("{0:20}{1:20}".format(word,le.lemmatize(word, get_wordnet_pos(word))))

--Word--            --Stem--            
plays               play                
played              played              
play                play                
playing             play                


In [22]:
def lemmatization(dataset):
    global le

    ## start ##
    new_dataset = [' '.join([le.lemmatize(word, get_wordnet_pos(word)) for word in word_tokenize(sentence)]) for sentence in dataset]
    ### end ###

    return new_dataset

print("Before lemmatization: ", test_sentence)
print("After lemmatization: ", lemmatization(test_sentence))

Before lemmatization:  ['I played and started playing with players in the playgrounds', 'I am playing with my playmates']
After lemmatization:  ['I played and start play with player in the playground', 'I be play with my playmate']


In [23]:
dataset_lemmatization = lemmatization(dataset)
vocab_lemmatization = construct_vocab(dataset_lemmatization)

vocab_analysis(vocab_lemmatization)

Vocab size:  17101
Vocab sample:  [('the', 10113), ('rock', 37), ('be', 6022), ('destine', 8), ('to', 4234), ('21st', 6), ('century', 22), ("'s", 3537), ('new', 206), ('``', 657)]


### 5. Get text processing methods what you want

In [24]:
def text_preprocessing(dataset):
    ## start ##
    dataset = lowercase(dataset)
    dataset = remove_stopwords(dataset)
    dataset = lemmatization(dataset)
    ### end ###

    return dataset

In [25]:
dataset_preprocessed = text_preprocessing(dataset)
vocab = construct_vocab(dataset_preprocessed)

vocab_analysis(vocab)

Vocab size:  16997
Vocab sample:  [('rock', 37), ('destine', 8), ('21st', 6), ('century', 22), ("'s", 3537), ('new', 206), ('``', 657), ('conan', 2), ('go', 316), ('make', 803)]


# Tokenization

In [26]:
def tokenization(dataset, vocab):
    dataset_preprocessed = text_preprocessing(dataset)

    ## start ##
    dataset_tokenization = [[vocab[word] for word in word_tokenize(sentence) if word in vocab] for sentence in dataset_preprocessed]
    ### end ###

    return dataset_tokenization

example = text_preprocessing([dataset[0]])

print(example)
print([vocab[word] for word in example[0].split()])

["rock destine 21st century 's new `` conan `` 's go make splash even great arnold schwarzenegger , jean-claud van damme steven segal ."]
[37, 8, 6, 22, 3537, 206, 657, 2, 657, 3537, 316, 803, 9, 387, 185, 15, 9, 10037, 1, 13, 2, 24, 1, 14010]


In [27]:
dataset_tokenization = tokenization(dataset, vocab)

# Homework: TF-IDF

### TF-IDF 구현
1. unique_words 생성 (!= vocab)
2. tf 구현
3. df 구현
4. idf 구현
5. tf, idf를 바탕으로 tf-idf 계산

### 평가 기준
1. 각 함수 작성 여부 - 60
2. 'chicago' 키워드로 검색했을 때 결과 - 20
    - text_preprocessing에 따라 결과가 다를 수 있음
3. 코드 실행 - 10
4. 보고서 - 10

In [28]:
import math
import pandas as pd
from collections import Counter
import numpy as np
from tqdm import tqdm
from typing import Dict, List


dataset_preprocessed = text_preprocessing(dataset)
vocab = construct_vocab(dataset_preprocessed)

In [29]:
dataset_preprocessed[0]

"rock destine 21st century 's new `` conan `` 's go make splash even great arnold schwarzenegger , jean-claud van damme steven segal ."

In [30]:
# 전체 단어 집합 만들기
unique_words = vocab.keys()

In [31]:
list(unique_words)[:10]

['rock',
 'destine',
 '21st',
 'century',
 "'s",
 'new',
 '``',
 'conan',
 'go',
 'make']

In [32]:
def term_frequency(data, unique_words) -> Dict[str, float]: # str: 문장 내 단어, float: 문장 내에서 단어 등장 비율
    tf = None

    ## start ##
    word_counts = Counter(word_tokenize(data))
    total_words = sum(word_counts.values())
    tf = {word: word_counts[word] / total_words for word in unique_words}
    ### end ###

    return tf

In [33]:
def document_frequency(dataset) -> Dict[str, int]: # str: 전체 단어, int: dataset 내에서 단어 등장 빈도
    df_counts = None

    ## start ##
    df_counts = Counter()
    for sentence in dataset:
        unique_tokens = set(word_tokenize(sentence))
        for word in unique_tokens:                   
            df_counts[word] += 1
    ### end ###

    return df_counts

In [34]:
def inverse_document_frequency(dataset, unique_words) -> Dict[str, float]: # str: 전체 단어, float: 역문서 빈도
    idf = None

    ## start ##
    N = len(dataset)
    df = document_frequency(dataset)
    idf = {word: math.log((N + 1) / (df.get(word, 1) + 1)) + 1 for word in unique_words}
    ### end ###

    return idf

In [35]:
def compute_tf_idf(dataset, unique_words, idf) -> List[Dict[str, float]]: # str: 문장 내 단어, float: tf-idf 값
    tf_idf_matrix = None

    ## start ##
    tf_idf_matrix = []
    for sentence in dataset:
        tf = term_frequency(sentence, unique_words)
        tf_idf_matrix.append({word: tf.get(word, 0) * idf[word] for word in unique_words})
    ### end ###

    return tf_idf_matrix

In [36]:
idf = inverse_document_frequency(dataset_preprocessed, unique_words)
tf_idf_matrix = compute_tf_idf(dataset_preprocessed, unique_words, idf)

In [37]:
# DataFrame으로 변환
tf_idf_df = pd.DataFrame(tf_idf_matrix).fillna(0)

# 결과 출력
tf_idf_df

,rock,destine,21st,century,'s,new,``,conan,go,make,...,stevenon,pamela,roller,coaster,re-working,chore,'time,waster,1959,made-for-home-video
0,0.27654,0.336555,0.347026,0.29746,0.190330,0.206722,0.387845,0.38233,0.189896,0.150849,...,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.000000,0.000000,0.000000
1,0.00000,0.000000,0.000000,0.00000,0.157514,0.000000,0.320975,0.00000,0.000000,0.000000,...,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.000000,0.000000,0.000000
2,0.00000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,...,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.000000,0.000000,0.000000
3,0.00000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.00000,0.414319,0.000000,...,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.000000,0.000000,0.000000
4,0.00000,0.000000,0.000000,0.00000,0.152264,0.000000,0.000000,0.00000,0.000000,0.000000,...,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.000000,0.000000,0.000000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10657,0.00000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,...,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.000000,0.000000,0.000000
10658,0.00000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,...,0.0,0.0,0.0,0.0,0.0,0.0,0.958139,0.958139,0.000000,0.000000
10659,0.00000,0.000000,0.000000,0.00000,0.084591,0.000000,0.000000,0.00000,0.000000,0.000000,...,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.000000,0.354866,0.000000
10660,0.00000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,...,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.000000,0.000000,1.596898


In [38]:
# 특정 입력과 tf_idf dot product score가 가장 높은 리뷰 10개 출력
def get_most_similar_reviews(input, tf_idf_df, idf, top_n=10):
    input_preprocess = text_preprocessing([input])

    input_tf_idf = compute_tf_idf(input_preprocess, unique_words, idf)[0]
    input_tf_idf_df = pd.DataFrame([input_tf_idf]).fillna(0)

    dot_products = tf_idf_df.dot(input_tf_idf_df.iloc[0])
    most_similar_reviews = dot_products.sort_values(ascending=False).head(top_n)
    most_similar_reviews = [(i, dataset[i], score) for i, score in most_similar_reviews.items()]

    for i, review, score in most_similar_reviews:
        print(f'{i}\t{score}\t{review}')

In [39]:
input_text = 'chicago'

get_most_similar_reviews(input_text, tf_idf_df, idf, top_n=10)

1019	7.755181839461603	chicago is , in many ways , an admirable achievement .
8008	5.6401322468811665	i'd give real money to see the perpetrators of chicago torn apart by dingoes .
100	5.6401322468811665	chicago is sophisticated , brash , sardonic , completely joyful in its execution .
4646	4.136096981046188	a stylish cast and some clever scripting solutions help chicago make the transition from stage to screen with considerable appeal intact .
5291	4.136096981046188	this chicago has hugely imaginative and successful casting to its great credit , as well as one terrific score and attitude to spare .
3374	3.6494973362172245	chicago offers much colorful eye candy , including the spectacle of gere in his dancing shoes , hoofing and crooning with the best of them .
9864	3.4467474842051566	the film didn't convince me that calvin jr . 's barbershop represents some sort of beacon of hope in the middle of chicago's south side .
7557	2.6974545528562093	the final result makes for adequate entert

## Introduction
### 1. 연구 목적 및 배경

자연어 처리(NLP) 분야에서는 문서 간 유사도를 정량적으로 측정하는 기술이 매우 중요하다.   
특히, 영화 리뷰, 상품 리뷰 등 사용자 의견이 담긴 텍스트에서 비슷한 감성을 가진 문장을 찾는 기술은 추천 시스템, 검색 엔진 등에 활용된다.

이번 과제에서는 Cornell Movie Review Dataset을 이용하여, 다음과 같은 시스템을 구현하였다:

- 텍스트 전처리 (소문자화, 불용어 제거, 표제어 추출 등)
- 단어 기반 TF-IDF 벡터화
- 사용자가 입력한 문장과 유사도가 가장 높은 리뷰 10개 출력

이를 통해, 기본적인 정보 검색 시스템의 핵심 요소인 TF-IDF 기반 벡터화와 유사도 연산을 실습하고,  
자연어 텍스트를 수치화하여 검색 기능으로 활용하는 전 과정을 구현하였다.


## Methodology
### 1. 데이터셋 개요

Cornell University에서 제공하는 Movie Review Polarity Dataset은 IMDB, Rotten Tomatoes와 같은 출처로부터 수집된 영화 리뷰 데이터로 구성된다.

- 긍정 리뷰: 5,331개 (rt-polarity.pos)
- 부정 리뷰: 5,331개 (rt-polarity.neg)
- 전체 샘플 수: 10,662개

해당 리뷰는 문장 단위의 짧은 리뷰로 구성되어 있으며, 텍스트 기반 감정 정보 또는 주제 분석에 적합하다.

### 2. 텍스트 전처리 과정

리뷰 데이터에 대해 다음과 같은 전처리를 단계적으로 수행하였다.

| 단계 | 설명 | 변화 | 
|-------|--------------|----------------|
| Lowercasing    | 모든 텍스트를 소문자로 통일       | The Rock is Destined → the rock is destined     |
| Stopword Removal     | 의미 없는 관사/전치사 제거       | the rock is destined to be → rock destined       |
| Lemmatization     | 단어를 원형으로 복원	      | players → player, playing → play         |

- nltk.word_tokenize()를 사용하여 단어 수준 토큰화를 일관성 있게 적용
- Stemming은 실험적으로 시도하였으나, 문법 파괴가 심해 최종적으로는 Lemmatization을 채택

### 3. Vocabulary 생성

- 전체 문서에 등장한 단어를 기준으로 vocab 사전 생성
- 전처리 전 단어 수: 20,298개
- 전처리 후 단어 수: 16,997개  
   → 약 16%의 노이즈 제거

### 4. TF-IDF 구현

TF-IDF는 다음 요소들로 구성되며, 직접 구현하였다.

- Term Frequency (TF): 문장 내 단어의 상대적 등장 비율
- Document Frequency (DF): 전체 문서에서 단어가 등장한 문서 수
- Inverse Document Frequency (IDF):  
  idf = log((N+1)/(df+1)) + 1 공식으로 smoothing 적용

`idf['rock'] = 1.94  # 예시`
- 각 문장마다 {단어: tf-idf score} 형태의 사전을 생성하고,
- pandas.DataFrame을 통해 10662 × 16997 크기의 희소 행렬 완성



## Results & Analysis

### 1. 전처리 결과 예시
`dataset_preprocessed[0]`  

→ "rock destine 21st century 's new `` conan `` 's go make splash even great arnold schwarzenegger , jean-claud van damme steven segal ."

- 의미 있는 핵심 단어만 추출되었고, stopwords 및 불필요한 구두점 제거 완료
- TF-IDF 기반 벡터화에 적합한 형식

`Initial Evaluation: {'eval_loss': 0.7041, 'eval_accuracy': 0.4909}`

- 초기 평가 정확도(Accuracy): 49.09%
- 손실 함수 값(Loss): 0.7041  

### 2. TF-IDF 행렬 분석

`tf_idf_df.head()`

- TF-IDF 값이 높은 단어: 문서 내 상대적으로 많이 등장하면서 전체 문서에서는 드문 단어
- 전체 리뷰 중 "conan"과 같은 고유명사나 "splash" 같은 강조 단어들이 강조됨

### 3. 유사도 검색 결과

입력 텍스트: "chicago"  
→ TF-IDF 기반 dot product를 활용하여 유사도 계산

출력 상위 3개 리뷰 예시

| 순위  | review index | 유사도 점수    | 내용     |
|-------|--------------|----------------|----------|
| 1     |    1019      | 7.76	        | chicago is , in many ways , an admirable achievement .   |
| 2     | 8008	       | 5.64           | i'd give real money to see the perpetrators of chicago torn apart by dingoes .   |
| 3     | 100          | 5.64           | chicago is sophisticated , brash , sardonic ...   |

- → “chicago”가 중심 키워드인 리뷰들이 정확히 상위로 랭킹됨
- → 전처리와 TF-IDF 벡터화의 정밀함 확인 가능


## Conclusion

### 1. 성과 요약

이번 과제에서는 영화 리뷰 데이터셋에 대해 TF-IDF 기반 유사 문장 검색 시스템을 성공적으로 구현하였다.

- 텍스트 전처리를 통해 의미 중심 단어 추출
- 직접 구현한 TF-IDF로 벡터화 수행
- dot product를 활용한 유사도 계산 → 정확한 검색 결과 도출

특히, word_tokenize, stopwords, lemmatizer, Counter 기반 TF, DF, IDF 계산 등 NLP 기초 기술들을 종합적으로 활용함으로써,  
TF-IDF 기반 정보 검색 시스템을 기초부터 구현하는 역량을 입증하였다.

### 2. 한계점 및 개선 방향

|                         한계                       |                         개선 방향                          |
|-------------------------------------|--------------|
| 단순한 TF-IDF 기반 벡터 → 문맥 정보 없음           |    BERT 임베딩, SBERT 기반 문장 벡터 고려 가능             |
| 단어 순서 및 문맥 손실	                         | RNN, Transformer 기반 문장 인코더 사용	                  |
| 정제된 단어 기준이므로 신조어, 오탈자 민감	     | Subword tokenizer (e.g. WordPiece, BPE) 활용 가능          |
